### PR TITLE
Fix argument handling in superdwarf

### PR DESCRIPTION
**Pull request by oorzkws**
There's a bit of cleaning I'm yet to do, I'll mark the PR as ready once it's done

### CODE REVIEW

**Comment by oorzkws** [comment]
I ended up redoing the internal architecture a bit to make command handling more straightforward.

This may be a base dfhack/structures issue, but this doesn't appear to work for smoothing tasks either on this version or the previous ones.

**Comment by myk002** [comment]
I'm surprised that smoothing isn't sped up like digging tasks are. I'm not sure what needs to be changed to get it to work. That might require some reverse engineering. Could you create a github issue for that so we (or you?) can look into it later?

**Comment by oorzkws** [comment]
Sorry for the delay, I was doing the edits in the web UI before and thus couldn't properly amend the PR.

It's now changed to use `superdwarf all` and `superdwarf clear`, the former being changed to keep the verbiage concise.

File: docs/superdwarf.rst [hunk 1]
```diff
@@ -18,6 +18,8 @@ Usage
     Give superspeed to selected creature.
 ``superdwarf add <ID>``
     Give superspeed to a creature by its unit ID.
+``superdwarf all``
+    Give superspeed to all of your citizens.
```
[the remark this answers is not in the document]
Comment (myk002) [reply docs/superdwarf.rst:22]: actually, scratch that. a player *could* have added another, non-citizen unit directly that `clear` will remove.

[the remark this answers is not in the document]
Comment (oorzkws) [reply docs/superdwarf.rst:22]: > actually, scratch that. a player _could_ have added another, non-citizen unit directly that `clear` will remove.

Exactly! Not sure if there's a better way to document this.

File: superdwarf.lua [hunk 2]
```diff
@@ -2,90 +2,113 @@
...
     end
+}
 
-    if argCommand == 'add' then
-        AddSuperdwarf(unit)
-        print("Applying superdwarf to [" .. unit.id .. "] " .. dfhack.TranslateName(unit.name))
-        return
-    end
+function main(...)
+    local args = ...
```
[the remark this answers is not in the document]
Comment (oorzkws) [reply superdwarf.lua:104]: Is main always called with a single table?

File: superdwarf.lua [hunk 3]
```diff
@@ -2,90 +2,113 @@
...
+end
 
-    if not unit then
-        return
+commands = {
+    add = function(arg)
+        local unit = getUnit(arg)
+        if not unit then
+            qerror('No unit found to add; select a unit or provide a unit id')
+            return
```
[the remark this answers is not in the document]
Comment (oorzkws) [reply superdwarf.lua:52]: Thanks, changed locally. I'll push once the other reviews are handled.

File: superdwarf.lua [hunk 4]
```diff
@@ -2,90 +2,113 @@
...
-
-function DeleteSuperdwarf(unit)
-    if superdwarfIds[unit.id] then
-        superdwarfIds[unit.id] = nil
-    else
-        qerror("Dwarf was not already super")
+    -- Passed an ID as a string or number
+    if type(obj) == 'string' or type(obj) == 'number' then
+        obj = tonumber(obj) -- Can be nil if someone passed something like a name
+        return obj and df.unit.find(tonumber(obj))
```
[the remark this answers is not in the document]
Comment (myk002) [reply superdwarf.lua:17]: `obj` is already set to `tonumber(obj)` in the previous line. if it is called with `'names'`, `obj' will already be `nil`

File: superdwarf.lua (hunk 2 above)
[the remark this answers is not in the document]
Comment (myk002) [reply superdwarf.lua:104]: according to line 115, yes, I think it is. Are there any other callers that would be different?

File: superdwarf.lua (hunk 4 above)
[the remark this answers is not in the document]
Comment (oorzkws) [reply superdwarf.lua:17]: Hm I may not be understanding here. If I did something like `superdwarf add urist` expecting it to work, this line would be necessary, no?

[the remark this answers is not in the document]
Comment (myk002) [reply superdwarf.lua:17]: Are we talking about the same line? This is what I was proposing:
```lua
obj = tonumber(obj)
return obj and df.unit.find(obj)
```

[the remark this answers is not in the document]
Comment (oorzkws) [reply superdwarf.lua:17]: Oh dear I completely missed the bottom line, I was focused on the middle portion. I'll fix that.

**Review by myk002 — APPROVED** [review]

**Pull request merged**